PullRequest for Issue2355: add generic beam on/off action to the CLSBeamlineStatusView and CLSShutters/CLSValves

XuetaoLiu-CLS opened the pull request:
#2355

dretrex commented on source/beamline/CLS/CLSShutters.cpp:
```diff
@@ -123,20 +123,36 @@ bool CLSShutters::isClosed() const
+		return safetyShutter_->isConnected() && isChildState1(safetyShutter_);
+	}
+
+	return true;
```
I think that it is probably better to return false as a default for whether this kind of equipment is open.

dretrex commented on source/beamline/CLS/CLSShutters.cpp:
```diff
@@ -123,20 +123,36 @@ bool CLSShutters::isClosed() const
 }
 
-bool CLSShutters::addShutter(AMControl *newShutter, double openValue, double closedValue, int beamOnOrder)
+bool CLSShutters::addShutter(AMControl *newShutter, int beamOnOrder, double openValue, double closedValue)
```
Have many people used this function?  If yes, you'll want to make sure you've gone over every implementation to ensure that it's being added correctly.

dretrex commented on source/beamline/CLS/CLSValves.cpp:
```diff
@@ -93,7 +93,7 @@ QList<AMControl*> CLSValves::closedValvesList() const
 }
 
-bool CLSValves::addValve(AMControl *newValve, double openValue, double closedValue, int beamOnOrder)
+bool CLSValves::addValve(AMControl *newValve, int beamOnOrder, double openValue, double closedValue)
```
Same with this.

dretrex commented:
Oh, woops.  I didn't bother looking at the fact that this wasn't ready for review.  Sorry about that @XuetaoLiu-CLS!

XuetaoLiu-CLS commented:
I just tested on SXRMB, the beam on/off is working as expected. By far, the generic beam on/off is only used by SXRMB. I will try to bring it to the other beamlines (Vespers and Ideas) later

@helfrij can you review this? Thx

helfrij commented on source/application/BioXAS/BioXASAppController.cpp:
```diff
@@ -204,22 +205,6 @@ void BioXASAppController::onRegionOfInterestBoundingRangeChanged(AMRegionOfInter
 		genericConfiguration_->setRegionOfInterestBoundingRange(region);
 }
 
-void BioXASAppController::goToBeamlineStatusView(AMControl *control)
```
Just curious--why is this functionality removed?

XuetaoLiu-CLS replied:
it is moved to CLSAppController

helfrij commented on source/application/BioXAS/BioXASAppController.cpp:
```diff
@@ -389,7 +374,7 @@ void BioXASAppController::createGeneralPanes()
 	addMainWindowView( beamlineConfigurationView, "Configuration", generalPaneCategeryName_, generalPaneIcon_);
 
-	beamlineStatusView_ = new CLSBeamlineStatusView(BioXASBeamline::bioXAS()->beamStatus(), false);
+	beamlineStatusView_ = new CLSBeamlineStatusView(BioXASBeamline::bioXAS(), false);
```
It looks like the declaration of the `beamlineStatusView_` was removed from the .h file, but we still create an instance of the view here. Does this cause a crash?

XuetaoLiu-CLS replied:
it is defined in the CLSAppController. This instantiation here might be able to moved to CLSAppController when we are having more beamlines using the CLSBeamStatusView

helfrij replied:
Ah okay.

helfrij commented on source/application/CLS/CLSAppController.h:
```diff
@@ -30,6 +32,9 @@ protected slots:
 	/// implementation for slot that connects generic scan editors that use the 2D scan view to the app controller so that it can enable quick configuration of scans.
 	virtual void onScanEditorCreatedImplementation(AMGenericScanEditor *editor);
 
+	/// Sets the beam status view as the current view, with the given control as the selected control.
```
Ah, I see. The method was moved up to CLSAppController.

helfrij commented on source/beamline/AMPseudoMotorControl.cpp:
```diff
@@ -100,8 +100,7 @@ bool AMPseudoMotorControl::validSetpoint(double value) const
 bool AMPseudoMotorControl::addChildControl(AMControl *control)
 {
-	if(AMControl::addChildControl(control)){
-		connect( control, SIGNAL(connected(bool)), this, SLOT(updateStates()) );
```
Does this connect statement appear elsewhere, so it doesn't need to be here?

XuetaoLiu-CLS replied:
`AMPseudoMotorControl` is a subclass of `AMConnectedControl`. Here I changed to use `AMConnectedControl::addChildChontrol(control)`, which should be a proper usage and also did the connected() signal/slot connection

helfrij replied:
Sounds good.

helfrij commented on source/beamline/BioXAS/BioXASBeamline.cpp:
```diff
@@ -513,19 +513,6 @@ bool BioXASBeamline::clearGe32Detectors()
 	return true;
 }
 
-void BioXASBeamline::setConnected(bool isConnected)
```
Is this functionality replaced elsewhere?

XuetaoLiu-CLS replied:
`setConnected()` was called by `updateConnected()` slot. And `updateConnected()` was replaced with `onBeamlineComponentConnected()` slot in the `CLSBeamline`

helfrij commented on source/beamline/BioXAS/BioXASBeamline.h:
```diff
@@ -418,14 +408,9 @@ protected slots:
 
 	/// Protected constructor.
-	BioXASBeamline(const QString &controlName);
+	BioXASBeamline(const QString &beamlineName, const QString &controlName);
```
What is the difference between the beamline name and the control name?

XuetaoLiu-CLS replied:
`beamlineName` is like a short version of the name, `controlName` is like a longer version, more detailed information.

This is primarily used for the messaging purpose

helfrij replied:
Oh okay. What's messaging?

dretrex replied:
I think he means when being used in warning message boxes or in the status view.

helfrij replied:
👍

helfrij commented on source/beamline/BioXAS/BioXASFrontEndShutters.cpp:
```diff
@@ -12,13 +12,12 @@ BioXASFrontEndShutters:: BioXASFrontEndShutters(const QString &name, QObject *pa
 
 	setUpstreamPhotonShutter(new AMReadOnlyPVControl("IPSH1407-I00-01", "IPSH1407-I00-01:state", this));
 	setDownstreamPhotonShutter(new CLSExclusiveStatesControl("IPSH1407-I00-02", "IPSH1407-I00-02:state", "IPSH1407-I00-02:opr:open", "IPSH1407-I00-02:opr:close", this));
-	setSafetyShutter(new CLSExclusiveStatesControl("SSH1407-I00-01", "SSH1407-I00-01:state", "SSH1407-I00-01:opr:open", "SSH1407-I00-01:opr:close", this));
```
Do we manage the safety shutter elsewhere, instead of with the other front-end shutters?

XuetaoLiu-CLS replied:
it is still here. I added two parameters to the `setSafetyShutter()` function.

XuetaoLiu-CLS replied:
@helfrij the implementation of `setSafetyShutter()` is moved to `CLSShutters`

helfrij replied:
Ah, okay.

helfrij commented on source/beamline/CLS/CLSBeamline.cpp:
```diff
@@ -1,12 +1,227 @@
+
+/// ==================== protected methods =======================
+
+void CLSBeamline::createBeamlineStatus(CLSShutters *shutters, CLSValves *valves)
```
Just curious--will the beamline status only ever consider shutters and valves?

XuetaoLiu-CLS replied:
no. it has shutters/valves/morror mask/monoMask as BioXAS has before.

Shutters and Valves are two common components cross beamlines. For convenience, we pass in shutters and valves here. at Ln306/ln307 of BioXASSideBeamline.cpp, you can see we added the MirrorMask and MonoMask

helfrij replied:
It looks like the ring current could also be incorporated eventually, based on the update beam status method above.

XuetaoLiu-CLS replied:
@helfrij maybe. My purpose here is that for the CLSBeamline, it will try to have some CLS level generalization. Eventually, all the beamline status will be determined by the ring current and the status of the  shutters/valves. 

(at this moment, each beamline has its own implementation, but similar to each other, for that)

helfrij commented on source/beamline/CLS/CLSBeamlineStatus.cpp:
```diff
@@ -10,18 +13,20 @@ CLSBeamlineStatus::CLSBeamlineStatus(const QString &name, QObject *parent) :
+	addOption(On, "Beam is ON", true);
+	addOption(Off, "Beam is OFF", true);
+
+	connect(this, SIGNAL(connected(bool)), this, SLOT(updateBeamlineStatus()));
```
I'm always a little leery about connecting our own signals to our own slots. Is it possible to make the calls to `updateBeamlineStatus()` in the places where these signals are originally emitted?

XuetaoLiu-CLS replied:
basically, CLSBeamlineStatus is a group of controls. The signals (`connected()`/`valueChanged()`) of any control managed by CLSBeamlineStatus will trigger CLSBeamlineStatus to emit the same signal.

In this way, connect to our own signals will avoid the mistake of missing a signal from the child controls.

dretrex replied:
@helfrij is there something in particular you don't like?  As long as there is no mechanism for signal-slot loops then this should be fine.

helfrij replied:
Given the option, I would prefer to reimplement the `AMEnumeratedControl::updateStates()` or `updateValue()` or something, and add whatever additional functionality is needed there (emitting the new `beamlineStatusChanged(bool)` signal).

dretrex replied:
Ah, gotcha.

helfrij commented on source/beamline/CLS/CLSBeamlineStatus.cpp:
```diff
@@ -10,18 +13,20 @@ CLSBeamlineStatus::CLSBeamlineStatus(const QString &name, QObject *parent) :
-	valvesControlSet_->clear();
-	mirrorMaskControlSet_->clear();
-	monoMaskControlSet_->clear();
+	clearComponents();
```
Does `clearComponents()` handle deleting components?

XuetaoLiu-CLS replied:
no. It only clears the control set.

helfrij replied:
That probably makes sense, as the CLSBeamlineStatus likely isn't the parent of each component.

helfrij commented on source/beamline/CLS/CLSShutters.cpp:
```diff
@@ -1,11 +1,16 @@
 CLSShutters::CLSShutters(const QString &name, QObject *parent) :
 	CLSTriStateGroup(name, parent)
 {
+	safetyShutter_ = 0;
```
Does the safety shutter belong here, conceptually? Must all groups of shutters have a safety shutter?

XuetaoLiu-CLS replied:
No. The saftyShutter is not mandatory.

dretrex replied:
To put this in context, is this meant to be the shutter that is not controlled by beamlines or the one that is?

XuetaoLiu-CLS replied:
@dretrex my understanding for this is that it is up to the user to make the decision how to use the safety shutter for CLSShutters.
For SXRMB:
The safety shutter is a Readonly PV control, which can't be turned on grammatically. The users complained several times that they can't turn beam on with acquaman. Acutally, the safety shutter is closed, but they don't know. With this changes, we will visually display the status of the safety shutter and also pop a message to tell them if the safety shutter is closed when they are trying to turn beam on.

For VESPERS, since we are able to open the safety shutter, we don't have to set the safety shutter, which can be identified as a regular shutter. If we still want to distinguish it with the other shutters, it will still works as I mentioned before.

To address @helfrij's question, safety shutter is not mandantory to the `CLSShutters`. You can use `CLSShutters` as what you did with the original `BioXASShutters`. It can be simply a group of shutters you want.

dretrex replied:
In that case, we should change that to not be safety shutter, but rather something like `addReadOnlyComponent` where the readonly components are important for the action to take place but cannot be done programmatically.

helfrij commented on source/beamline/CLS/CLSShutters.h:
```diff
@@ -60,6 +82,13 @@ public slots:
+	AMControl *safetyShutter_;
+
+	/// the order of the shutters in the beam on action. If a shutter is not invovled in the beam on/off action, the order is -1
+	QMap<int, AMControl*> shuttersBeamOnOrderMap_;
```
I like the inclusion of an order. Nice!

helfrij commented on source/beamline/CLS/CLSShutters.cpp:
```diff
@@ -14,7 +19,88 @@ CLSShutters::CLSShutters(const QString &name, QObject *parent) :
+	// this is to make sure all the controls are checked
+	int currentBeamOffOrder = 1;
+	int checkedControlCount = 0;
+	while (checkedControlCount < shuttersBeamOnOrderMap_.count()) {
```
Just checking: shutters are opened in order of map key, and closed in the order in which they were added to the map?

XuetaoLiu-CLS replied:
it is still in order of map keys. The next line of codes gets the shutter in the order. This `while` loop will make sure we go through all the ordered shutters

helfrij replied:
Oh yes, I see. I misinterpreted what was going on.

dretrex replied:
Shouldn't the turn off order be the inverse of the turn on order?

helfrij commented on source/stylesheets/AMToolButton.qss:
```diff
@@ -53,7 +53,7 @@ background-color: #800000;
 
 AMToolButton[colorState="3"] {
-background-color: qlineargradient(x1:0 y1:0, x2:0 y2:1, stop:0 #ffff00, stop:1 #808000);
+ background-color: qlineargradient(x1:0 y1:0, x2:0 y2:1, stop:0 #ffffff, stop:1 #808080);
```
Could this change be handled in #2449 instead?

helfrij commented on source/ui/CLS/CLSValueEditor.cpp:
```diff
@@ -55,6 +55,11 @@ CLSValueEditor::~CLSValueEditor()
 
 }
 
+void CLSValueEditor::addLayout(QLayout *newLayout)
```
Is having this functionality necessary? I can see it being a source of complexity/confusion down the road.

XuetaoLiu-CLS replied:
nice catch. I forgot why I added this, but looks like it is never used. I will remove it

helfrij commented on source/beamline/CLS/CLSBeamlineStatus.cpp:
```diff
@@ -33,11 +38,7 @@ bool CLSBeamlineStatus::isOn() const
+	return isConnected() && !isOn();
 }
 
 AMControl * CLSBeamlineStatus::beamlineStatusPVControl()
```
I don't think BioXAS does, and I'm just curious--do many beamlines have a beamline status PV control like this?

XuetaoLiu-CLS replied:
SXRMB has one. BioXAS doesn't. This is optional.

dretrex replied:
VESPERS kind of does.  It's all over the map.

helfrij commented on source/beamline/CLS/CLSShutters.cpp:
```diff
@@ -14,7 +19,88 @@ CLSShutters::CLSShutters(const QString &name, QObject *parent) :
+	shuttersBeamOnOrderMap_.clear();
+}
+
+AMListAction3* CLSShutters::createBeamOnActionList()
```
Is there something beam on/off specific about these action lists? Or are they just opening/closing the shutters and valves? If it's the latter case, we could think about them instead in terms of a shutters move action (eg. `moveToOpen()`, `moveToClosed()`) and reimplement `AMPseudoMotorControl::createMoveAction(double)`.

XuetaoLiu-CLS replied:
it is just opening/closing the shutters/valves. and yes, it will be better to reimplement the `AMPseudoMotorControl::createMoveAction(double)`

helfrij replied:
I think your solution should probably replace the existing `moveToOpen/ClosedAction()` functionality.

dretrex replied:
Just to confirm: @helfrij, are you suggesting this code live inside `CLSShutters`?

helfrij replied:
Yes, I don't think it makes sense to move it out of CLSShutters. There is an existing implementation of `CLSShutters::createMoveToOpen/ClosedActions()` in this class that doesn't take the newly implemented order into account; I think @XuetaoLiu-CLS's code in `CLSShutters::createBeamOn/OffActionList()` could replace the older code.

helfrij commented:
A couple of questions.

Time Estimate: 1 hour.

XuetaoLiu-CLS commented:
@helfrij thanks for going through this giant PR. Lots of codes are changed. I addressed most of your comments. Please take a look.

I will try to use the `createMoveAction()` to replace the `createMoveOnActions()`

dretrex commented:
Hey @XuetaoLiu-CLS, once you replace the createMoveOnActions let me know and I'll go over it then!

XuetaoLiu-CLS commented:
@dretrex @helfrij I just checked the CLSShutters/CLSValves. There were an implementation of `createMoveAction()`, which will create a move action to open/close all the shutters/valves.

For beam on/off action list, we need to create the open/close actions in a given order, which might NOT include all the shutters/valves. 

I think it might be better to keep the current solution so far. What do you think?

dretrex commented on source/application/PGM/PGMAppController.cpp:
```diff
@@ -186,14 +171,14 @@ void PGMAppController::setupUserInterfaceImplementation()
 {
 	PGMPersistentView *pgmPersistentView = new PGMPersistentView;
-	connect( pgmPersistentView, SIGNAL(beamlineStatusSelectedComponentChanged(AMControl*)), this, SLOT(goToBeamlineStatusView(AMControl*)) );
+	connect( pgmPersistentView, SIGNAL(beamlineStatusControlClicked(AMControl*)), this, SLOT(goToBeamlineStatusView(AMControl*)) );
```
Was this connect statement not altered in BioXASAppController or was it done in a different pull request?

XuetaoLiu-CLS replied:
in PR for Issue2346 (#2428), @helfrij altered it to  `beamStatusControlClicked()`

dretrex commented on source/application/SXRMB/SXRMBAppController.cpp:
```diff
@@ -195,9 +189,9 @@ void SXRMBAppController::onBeamAvailabilityChanged(bool beamAvailable)
-	return AMChooseDataFolderDialog::getDataFolder("/AcquamanLocalData/sxrmb",  //local directory
-												   "/home/sxrmb",               //remote directory
-												   "acquamanData",              //data directory
+	return AMChooseDataFolderDialog::getDataFolder("/AcquamanLocalData",  //local directory
```
This seems like a random change.  What was the reason for it?

dretrex commented on source/beamline/BioXAS/BioXASBeamline.h:
```diff
@@ -66,7 +61,7 @@ class BioXASBeamline : public CLSBeamline
 	{
 		if (!instance_) {
-			instance_ = new BioXASBeamline("BioXAS Beamline");
+			instance_ = new BioXASBeamline("BioXASBeamline", "BioXAS Beamline");
```
Based on your answer to @helfrij just below, I'd suggest the name be `BioXAS` (other beamlines being the same).  Otherwise, this short name will only ever be one space shorter than the long name which adds no value.

dretrex commented on source/beamline/CLS/CLSBeamline.cpp:
```diff
@@ -1,12 +1,227 @@
+		return 0;
+	}
+
+	if (!beamlineShutters_->isSafetyShutterOpen()) {
```
Using VESPERS as an example, beam on would include opening a safety shutter and would therefore never make it past this statement.  Is there a means of bypassing that?

XuetaoLiu-CLS replied:
@dretrex good to know about that. This version of generic beam on/off only works for SXRMB. If the safety shutter can be turned on programmatically, we will only make this check if the safety shutter is NOT in the beam on mapping.

dretrex commented on source/beamline/CLS/CLSBeamlineStatus.cpp:
```diff
@@ -122,8 +134,15 @@ bool CLSBeamlineStatus::clearComponents()
 
-	if (result)
+	if (result) {
+		shuttersControlSet_->clear();
```
Use `clearComponents()` here?

XuetaoLiu-CLS replied:
this is the implementation of `clearComponents()`.

dretrex replied:
lol, so it is.

dretrex commented on source/beamline/CLS/CLSShutters.cpp:
```diff
@@ -14,7 +19,88 @@ CLSShutters::CLSShutters(const QString &name, QObject *parent) :
+	// this is to make sure all the controls are checked
+	int currentBeamOnOrder = 1;
+	int checkedControlCount = 0;
+	while (checkedControlCount < shuttersBeamOnOrderMap_.count()) {
```
Why don't you just grab all the keys from the map and then use a for loop to iterate through each of them?

XuetaoLiu-CLS replied:
@dretrex I want to make sure they are ordered. And yes, I can use that. Just checked the documents:
```
QList<Key> QMap::keys() const
Returns a list containing all the keys in the map in ascending order.
```

dretrex commented on source/beamline/CLS/CLSValves.h:
```diff
@@ -16,6 +23,9 @@ class CLSValves : public CLSTriStateGroup
 
+	/// helper function to create a beam on action list
+	AMListAction3* createBeamOnActionList();
+
```
No off?

helfrij replied:
That might be a safety thing. I'm not sure we'd want to turn off beam by attempting to close all the valves.

XuetaoLiu-CLS replied:
@dretrex we never turn Valves off when we turning beam off.
Also in the following implementation, it also mentioned that we don't support close valves
```
AMAction3* CLSValves::createMoveAction(double setpoint)
{
	AMAction3 *result = 0;

	if (setpoint == Open)
		result = createMoveToOpenAction();

	// Closing valves is not supported at this time.
	// This is because a user accidentally closing valves
	// could trigger machine protection protocols.

	return result;
}
```

dretrex replied:
I was thinking more as a general CLSValves aspect rather than being a coupled part of beam status.  Generally, the use case for closing valves is minimal but I didn't want us to not think about it.

dretrex commented on source/beamline/SXRMB/SXRMBBeamline.cpp:
```diff
@@ -670,32 +579,26 @@ void SXRMBBeamline::setupComponents()
-	beamlineValves_->addValve(VVR16064B1006Valve_, CLSBEAMLINE_VALVE_OPEN, CLSBEAMLINE_VALVE_CLOSED);
-	beamlineValves_->addValve(VVR16064B1007Valve_, CLSBEAMLINE_VALVE_OPEN, CLSBEAMLINE_VALVE_CLOSED);
-	beamlineValves_->addValve(VVR16065B1001Valve_, CLSBEAMLINE_VALVE_OPEN, CLSBEAMLINE_VALVE_CLOSED);
+	beamlineValves_->addValve(VVR16064B1003Valve, CLSExclusiveStatesControl::Open, CLSExclusiveStatesControl::Closed, 1);
```
Quick question about SXRMB automatic valve opening.  Were there any waits between the opening of consecutive valves?  If not, is this a feature they use often?  I know from past experience that valves are typically opened with a bit of time between valve to minimize pressure waves inside the vacuum chambers.

XuetaoLiu-CLS replied:
@dretrex there are waits actions. This is not a frequently used feature. Usually, the valves are all open. To put valves in the beam on actions is to make sure the beam will be ON if they want.

dretrex commented on source/ui/CLS/CLSBeamlineStatusView.h:
```diff
@@ -54,26 +62,43 @@ protected slots:
 	virtual QWidget* createComponentView(AMControl *control);
 
+	/// helper functions to create a push button
+	QPushButton *createPushButton(QString text);
```
`const QString &`.  Also, technically, a `const` function too.

helfrij commented on source/beamline/BioXAS/BioXASFrontEndShutters.cpp:
```diff
@@ -46,7 +45,7 @@ void BioXASFrontEndShutters::setUpstreamPhotonShutter(AMReadOnlyPVControl *newCo
 
 		if (upstreamPhotonShutter_)
-			addShutter(upstreamPhotonShutter_, 1, 4);
+			addShutter(upstreamPhotonShutter_, CLS_SHUTTER_OPEN, CLS_SHUTTER_CLOSED);
```
Quick question. Do these `CLS_SHUTTER_OPEN/CLOSED` values represent the PV open/closed, or the pseudomotor CLSShutters open/closed? Remember that those will be different values.

XuetaoLiu-CLS replied:
@helfrij yes, that confused me before. the Open/Close value we used in the addShutter function should be the one we used for the control.

That is, if the control is a regular PV, the open/close should be 1/4 as the case here. If the control is a pseudomotor, the open/close should be 0/1 (or 1/0) (this depends on the open/close definition of that pseudomotor).

XuetaoLiu-CLS replied:
@helfrij and also in this case, we were using the magic number (1, 4). I simply replaced it with the state definitions,

helfrij replied:
Thanks @XuetaoLiu-CLS!

dretrex commented:
Lots of code and a bunch of questions, but overall looking alright.
Time Estimate: 50 minutes

XuetaoLiu-CLS commented:
@dretrex @helfrij when going through the comments, I had a thought. It might be less confusing if I keep the `CLShutters/CLSValves` as it was. There should be a subclass, called `CLSBeamStatusShutters/CLSBeamStatusValves`, of CLShutters/CLSValves, which will take care of the safety shutters and beam on/off actions.

helfrij commented:
Lots of the changes, introducing the ordering and updating the move actions in particular, enhance the existing functionality of CLSShutters and CLSValves. If there are SXRMB-specific things you'd like to keep, perhaps they could go in an SXRMBShutters subclass?

dretrex commented:
To combine both ideas, I think it would make a lot of sense having the upgraded CLSValves/CLSShutters, but then making a CLSBeamStatus that has both of them.  There is no reason to make a beam status version of them.  Then we could still subclass the CLSValves/CLSShutters for specific beamline needs without compromising the beam status part.

XuetaoLiu-CLS commented:
@dretrex @helfrij I made the new changes to have a CLSBeamStatusShutters/CLSBeamStatusValves as a subclass of CLSShutters/CLSValves. Please take a look at the changes.

@dretrex I saw you comment. If you still think it will be better not to have the specified beamstatus shutters/valves after you reviewed the new changes, I can revert the changes.

helfrij commented on source/beamline/AMConnectedControl.h:
```diff
@@ -13,6 +13,11 @@ class AMConnectedControl : public AMControl
 	/// Destructor.
 	virtual ~AMConnectedControl();
 
+	/// Adds a given control to the list of child controls.
```
What's the reason for moving these methods from public to protected?

dretrex replied:
It's actually the opposite.  But a reason would still be appreciated.

dretrex commented on source/beamline/CLS/CLSValves.h:
```diff
@@ -13,8 +16,6 @@ class CLSValves : public CLSTriStateGroup
 	/// Constructor.
 	explicit CLSValves(const QString &name, QObject *parent = 0);
-	/// Destructor.
-	virtual ~CLSValves();
```
Why was this removed?

dretrex commented on source/ui/CLS/CLSBeamlineStatusView.h:
```diff
@@ -54,26 +62,43 @@ protected slots:
 	bool compactView_;
+
+	/// The beamline being operated
+	CLSBeamline *beamline_;
```
This is unnecessary because there will only ever be one beamline which can always be accessed by `CLSBeamline::bl()`.

dretrex commented:
Hey @XuetaoLiu-CLS, after looking over it, I think extra stage of beam status shutters and valves obfuscates what you're trying to achieve.  I do still think my suggestion is the better way to go on this issue.

helfrij commented on source/beamline/AMPseudoMotorControl.cpp:
```diff
@@ -112,7 +112,7 @@ bool AMPseudoMotorControl::addChildControl(AMControl *control)
 bool AMPseudoMotorControl::removeChildControl(AMControl *control)
 {
-	if (AMConnectedControl::removeChildControl(control)){
+	if (AMControl::removeChildControl(control)){
```
Why the change from AMConnectedControl to AMControl here?

helfrij commented on source/beamline/BioXAS/BioXASFrontEndShutters.cpp:
```diff
@@ -70,9 +70,9 @@ void BioXASFrontEndShutters::setDownstreamPhotonShutter(CLSExclusiveStatesContro
-	actionList->addSubAction(createCheckChildIsOpenAction(upstreamPhotonShutter_, 10));
-	actionList->addSubAction(createMoveChildToOpenAction(safetyShutter_));
-	actionList->addSubAction(createMoveChildToOpenAction(downstreamPhotonShutter_));
+	actionList->addSubAction(createCheckChildIsOpen(upstreamPhotonShutter_, 10));
```
Why the rename here? The 'action' part seems important, as that's what's actually returned.

XuetaoLiu-CLS commented:
@dretrex the changes for CLSBeamStatusShutters/CLSBeamStatusValves are reverted. The  CLSShutters/Valves was upgraded with the beam on/off feature; the CLSBeamlineStatus contains both CLSShutters/CLSValves and supports the beam on/off feature.
